Generic list component and refactoring

TheSlimvReal opened the pull request:
see issue: #512 #511 #199 

### Visible/Frontend Changes
- List component (Notes, Children, Schools) look and behave more similar

### Architectural/Backend Changes
- Generic entity list component which is used in Notes, Schools and Children
- School and Child Details are now replaced by generic Entity Details

sleidig commented on src/app/core/entity-components/entity-list/entity-list.component.ts:
```diff
@@ -0,0 +1,270 @@
+  filtersConfig: any[] = [];
+
+  ready = true;
+  columnsToDisplay: any[] = [];
```
```suggestion
  columnsToDisplay: string[] = [];
```

array of string column ids, isn't it?

sleidig commented on src/app/core/entity-components/entity-list/entity-list.component.ts:
```diff
@@ -0,0 +1,270 @@
+  @ViewChild(MatPaginator) paginator: MatPaginator;
+
+  listName = "";
+  columns: any[] = [];
```
let's use (or introduce) interfaces with an explicit typing for `columns` and `filtersConfig`. This makes it incredibly easier later on to work with it.

sleidig reviewed (changes requested):
functionality seems to work correctly.

Let's clean up code in a few places, then it should be good to go :+1:

TheSlimvReal commented on src/app/child-dev-project/children/children-list/list-attendance/list-attendance.component.ts:
```diff
@@ -0,0 +1,78 @@
+    ></app-attendance-block>
+  `,
+})
+export class ListAttendanceComponent implements OnInitDynamicComponent {
```
We could do that but on the other hand it would limit the components in their usage. The `ChildBlock` component is used in many locations, not just as a column.

sleidig replied:
Oh, I misinterpreted the use of this a bit, maybe. As I understand it now this component (loads and) displays the most recent attendance blocks for one child, right?

Then maybe naming it something like `RecentAttendanceBlocksComponent` or `RecentAttendanceBlocksOfChildComponent`?

This should also go into a jsdoc for the component.

TheSlimvReal commented on src/app/child-dev-project/educational-material/educational-material-component/educational-material.component.ts:
```diff
@@ -69,7 +69,7 @@ export class EducationalMaterialComponent
 
   onInitFromDynamicConfig(config: any) {
-    this.child = config.child;
+    this.child = config.entity;
```
That is true. On the other hand I did not want to push changes without the reason why they were introduced. But this PR is definitely to big. I hope we can get faster cycles in the next features.

TheSlimvReal commented on src/app/core/entity-components/display-date/display-date.component.ts:
```diff
@@ -0,0 +1,16 @@
+import { OnInitDynamicComponent } from "../../view/dynamic-components/on-init-dynamic-component.interface";
+
+@Component({
+  selector: "app-display-date",
```
We could start creating bundles with what can be used in the list and what in the details section, but some components might be used in either.

TheSlimvReal commented on src/app/core/entity-subrecord/entity-subrecord.module.ts:
```diff
@@ -1,44 +0,0 @@
-  declarations: [EntitySubrecordComponent, KeysPipe],
-  exports: [EntitySubrecordComponent, KeysPipe],
-})
-export class EntitySubrecordModule {}
```
So we create a separate module for every component?

sleidig replied:
Angular Material does exactly that (although importing all those modules can be a real hassle as well).

I don't have very strong opinions on this but generally having less pieces linked together makes things a lot easier to understand. And in our architecture terminology the EntityDetails and EntityList are "views" and the SubrecordComponent is a "section".

sleidig commented on src/app/child-dev-project/children/children-list/list-attendance/list-attendance.component.ts:
```diff
@@ -0,0 +1,78 @@
+  attendanceList: AttendanceMonth[] = [];
+  maxAttendanceBlocks: number = 3;
+
+  @Input() filterBy: string;
```
These inputs would also benefit from a jsdoc. and maybe naming it `filterByInstitution` would be more fitting?

TheSlimvReal commented:
Changes from last push:

- Introduced types for the configuration
- Placed `EntityList`, `EntityDetails` and `EntitySubrecord` into separate modules
- Fixed bug where `ConflictResolution` was always added to the navigation bar
- Created static functions which parse the route without query params -> Follow up: still many components use static routes (#553)
- Added jsdoc to the new components
- Minor fixes that were requested and more that were found in the process (e.g. directly displaying a note that just got added)

sleidig commented on src/app/child-dev-project/children/children-list/children-list.component.spec.ts:
```diff
@@ -19,22 +19,30 @@ import { ActivatedRoute, Router } from "@angular/router";
+      { component: "DisplayText", title: "Class", id: "schoolClass" },
+      { component: "DisplayText", title: "School", id: "schoolId" },
+      {
+        component: "ListAttendance",
```
didn't this component name change to "recent attendance blocks" ...?

sleidig commented on src/app/child-dev-project/children/children-list/recent-attendance-blocks/recent-attendance-blocks.component.ts:
```diff
@@ -52,10 +58,10 @@ export class ListAttendanceComponent implements OnInitDynamicComponent {
-  onInitFromDynamicConfig(config: any) {
-    this.filterBy = config.id;
+  onInitFromDynamicConfig(config: ColumnCellConfig) {
+    this.filterByInstitution = config.id;
```
should this really map to `config.id`?

sleidig commented on src/app/child-dev-project/notes/note-details/note-details.component.ts:
```diff
@@ -19,7 +20,14 @@ export class NoteDetailsComponent implements ShowsEntity, OnInit {
+  ) {
+    this.matDialogRef
+      .beforeClosed()
+      .subscribe(() => this.matDialogRef.close(this.entity));
```
What does this do? looks like hooking into the closing process to just manually close it again?

sleidig commented on src/app/core/view/dynamic-components-map.ts:
```diff
@@ -38,10 +38,10 @@ export const DYNAMIC_COMPONENTS_MAP = new Map<string, any>([
   ["ChildBlock", ChildBlockComponent],
   ["ChildBlockList", ChildBlockListComponent],
-  ["ListAttendance", ListAttendanceComponent],
+  ["ListAttendance", RecentAttendanceBlocksComponent],
```
Can we rename it here as well? 
```suggestion
  [" RecentAttendanceBlocks", RecentAttendanceBlocksComponent],
```

sleidig commented on src/app/child-dev-project/children/child-details/form/form.component.ts:
```diff
@@ -57,20 +53,22 @@ export class FormComponent implements OnChanges {
-      this.router.navigate(["/child", this.child.getId()]);
-      // }
+      await this.entityMapperService.save<Entity>(this.entity);
+      const route = this.entity.getConstructor().ENTITY_TYPE.toLowerCase();
```
see #553

sleidig commented on src/app/core/entity-components/entity-list/EntityListConfig.ts:
```diff
@@ -0,0 +1,52 @@
+  component: string;
+  title: string;
+  id: string;
+  // This config could be anything that the component understands to parse
```
```suggestion
  /** this config can be anything that the component understands to parse */
```
